lib/util.js

```diff
@@ -361,262 +371,229 @@ function formatValue(ctx, value, recurseTimes) {
 
   // Provide a hook for user-specified inspect functions.
   // Check that value is an object with an inspect function on it
-  if (ctx.customInspect && value) {
-    const maybeCustomInspect = value[customInspectSymbol] || value.inspect;
-
-    if (typeof maybeCustomInspect === 'function' &&
-        // Filter out the util module, its inspect function is special
-        maybeCustomInspect !== exports.inspect &&
-        // Also filter out any prototype objects using the circular check.
-        !(value.constructor && value.constructor.prototype === value)) {
-      let ret = maybeCustomInspect.call(value, recurseTimes, ctx);
-
-      // If the custom inspection method returned `this`, don't go into
-      // infinite recursion.
-      if (ret !== value) {
-        if (typeof ret !== 'string') {
-          ret = formatValue(ctx, ret, recurseTimes);
-        }
-        return ret;
+  const customInspect = isCustomInspect(ctx, value);
+  if (customInspect !== undefined) {
+    if (ctx.seen === undefined) {
+      ctx.seen = [];
```

Comment:
`ctx.seen = [...ctx.seenSet]`?

Also, I, personally, would be fine with having `seen` just be a getter/setter pair.